doc/do.md

```diff
@@ -348,6 +348,44 @@ TODO: describe `forIn`
 
 TODO
 
+## Returning early from a failed match
+
+Inside a `do` block, the pattern `let _ ← <success> | <fail>` will continue with the rest of the block if the match on the left hand side succeeds, but will execute the right hand side and exit the block on failure:
+
+```lean
+def showUserInfo (username favoriteColor : Option String) : IO Unit := do
+  let some n ← username | IO.println "no username"
+  IO.println s!"username: {n}"
+  let some c ← favoriteColor | IO.println "user didn't provide a favorite color"
+  IO.println s!"favorite color: {c}"
+
+-- username: JohnDoe
+-- favorite color: red
+#eval showUserInfo (some "JohnDoe") (some "red")
+
+-- no username
+#eval showUserInfo none (some "purple")
+
+-- username: JaneDoe
+-- user didn't provide a favorite color
+#eval showUserInfo (some "JaneDoe") none
+```
+
+## If-let
+
+Inside a `do` block, users can employ the `if let` pattern to destructure actions:
+
+```lean
+def tryLength (input : Option String) : Except String Nat := do
+  if let some x ← input then return x.length else Except.error "No input!"
```

Comment:
It's a bit odd to use `←` here since `input` is not monadic.

Reply:
You right. I changed it to `IO (Option Nat)`.

Reply:
The examples still *look* pretty pure; they should at least prefix monadic parameter names with "get" and use `pure <| some` throughout.

Reply:
@Kha 

You might have to clue me in on what the idiom is for `pure <| some`; the examples I used were supposed to mimic this: https://github.com/leanprover/lean4/blob/420f5bb315f7cf9b8ef628c0aea0049b43ae2402/src/Lean/Util/Path.lean#L42. I know there are more meaty examples in Meta/Elab, but they don't seem appropriate for a doc example covering syntax.

Reply:
I only meant that you should use `pure <| some` consistently in both example blocks, which in fact should be mandatory now.

Reply:
> The examples still _look_ pretty pure; they should at least prefix monadic parameter names with "get"

So, what do you think? What I had in mind was `getInput` etc.

Reply:
Sounds good.